flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/join/temporal/TemporalRowTimeJoinOperator.java

```diff
@@ -0,0 +1,437 @@
+ * is that between watermarks we are collecting those elements and once we are sure that there will be
+ * no updates we emit the correct result and clean up the expired data in state.
+ *
+ * <p>Cleaning up the state drops all of the "old" values from the probe side, where "old" is defined
+ * as older then the current watermark. Build side is also cleaned up in the similar fashion,
+ * we sort all "old" values with row time and row kind and then clean up the old values, when clean up
+ * the "old" values, if the latest record of all "old" values is retract message which means the version
+ * end, we clean all "old" values, if the the latest record is accumulate message which means the version
+ * start, we keep the latest one, and clear other "old" values.
+ *
+ * <p>One more trick is how the emitting results and cleaning up is triggered. It is achieved
+ * by registering timers for the keys. We could register a timer for every probe and build
+ * side element's event time (when watermark exceeds this timer, that's when we are emitting and/or
+ * cleaning up the state). However this would cause huge number of registered timers. For example
+ * with following evenTimes of probe records accumulated: {1, 2, 5, 8, 9}, if we
+ * had received Watermark(10), it would trigger 5 separate timers for the same key. To avoid that
+ * we always keep only one single registered timer for any given key, registered for the minimal
+ * value. Upon triggering it, we process all records with event times older then or equal to
+ * currentWatermark.
+ */
+public class TemporalRowTimeJoinOperator
+	extends BaseTwoInputStreamOperatorWithStateRetention {
+
+	private static final long serialVersionUID = 1L;
+
+	private static final String NEXT_LEFT_INDEX_STATE_NAME = "next-index";
+	private static final String LEFT_STATE_NAME = "left";
+	private static final String RIGHT_STATE_NAME = "right";
+	private static final String REGISTERED_TIMER_STATE_NAME = "timer";
+	private static final String TIMERS_STATE_NAME = "timers";
+
+	private final boolean isLeftOuterJoin;
+	private final InternalTypeInfo<RowData> leftType;
+	private final InternalTypeInfo<RowData> rightType;
+	private final GeneratedJoinCondition generatedJoinCondition;
+	private final int leftTimeAttribute;
+	private final int rightTimeAttribute;
+
+	/**
+	 * The comparator to get ordered elements of right state.
+	 */
+	private final ChangelogOrderComparator rightChangelogOrderComparator;
+
+	/**
+	 * Incremental index generator for {@link #leftState}'s keys.
+	 */
+	private transient ValueState<Long> nextLeftIndex;
+
+	/**
+	 * Mapping from artificial row index (generated by `nextLeftIndex`) into the left side `Row`.
+	 * We can not use List to accumulate Rows, because we need efficient deletes of the oldest rows.
+	 */
+	private transient MapState<Long, RowData> leftState;
+
+	/**
+	 * Mapping from timestamp to right side `Row`.
+	 **/
+	private transient MapState<RowData, Long> rightState;
```

Comment:
This can be `MapState<Long, RowData>` to have better performance as we discussed.